tests/test-snapd-desktop-integration.c

```diff
@@ -0,0 +1,352 @@
+{
+   g_assert_true(launch_snapd_desktop_integration());
+}
+
+static void
+handle_notifications_method_call(GDBusConnection *connection, const gchar *sender, const gchar *object_path, const gchar *interface_name, const gchar *method_name, GVariant *parameters, GDBusMethodInvocation *invocation, gpointer user_data)
+{
+   if (strcmp(interface_name, "org.freedesktop.Notifications") != 0) {
+      g_dbus_method_invocation_return_error(invocation, G_DBUS_ERROR, G_DBUS_ERROR_UNKNOWN_INTERFACE, "Unknown interface");
+      return;
+   }
+
+   if (strcmp(method_name, "GetServerInformation") == 0) {
+      g_dbus_method_invocation_return_value(invocation, g_variant_new("(ssss)", "mock-notifications-server", "Test", "1.0", "1.2"));
+   } else if (strcmp(method_name, "GetCapabilities") == 0) {
+      g_autoptr(GVariantBuilder) capabilities = g_variant_builder_new(G_VARIANT_TYPE("as"));
+      g_dbus_method_invocation_return_value(invocation, g_variant_new("(as)", capabilities));
+   } else if (strcmp(method_name, "Notify") == 0) {
+      const gchar *app_name, *app_icon, *summary, *body;
+      guint32 replaces_id;
+      g_auto(GStrv) actions;
+      g_autoptr(GVariantIter) hints = NULL;
+      gint32 expire_timeout;
+      g_variant_get(parameters, "(&su&s&s&s^asa{sv}i)", &app_name, &replaces_id, &app_icon, &summary, &body, &actions, &hints, &expire_timeout);
+      guint32 notification_id = next_notification_id;
+      next_notification_id++;
+      g_dbus_method_invocation_return_value(invocation, g_variant_new("(u)", notification_id));
+
+      if (state == STATE_PROMPT_INSTALL) {
+         g_assert_cmpstr(summary, ==, "Some required theme snaps are missing.");
+         g_assert_cmpstr(body, ==, "Would you like to install them now?");
+         g_assert_cmpint(g_strv_length(actions), ==, 4);
+         g_assert_cmpstr(actions[0], ==, "yes");
+         g_assert_cmpstr(actions[1], ==, "Yes");
+         g_assert_cmpstr(actions[2], ==, "no");
+         g_assert_cmpstr(actions[3], ==, "No");
+
+         g_assert_true(g_dbus_connection_emit_signal(connection, NULL, "/org/freedesktop/Notifications", "org.freedesktop.Notifications", "ActionInvoked", g_variant_new("(us)", notification_id, "yes"), NULL));
+
+         state = STATE_INSTALL_THEMES;
+      } else if (state == STATE_NOTIFY_COMPLETE) {
+         g_assert_cmpstr(summary, ==, "Installing missing theme snaps:");
+         g_assert_cmpstr(body, ==, "Complete.");
+         exit_code = EXIT_SUCCESS;
+         g_main_loop_quit(loop);
+      }
+   } else {
+      g_dbus_method_invocation_return_error(invocation, G_DBUS_ERROR, G_DBUS_ERROR_UNKNOWN_METHOD, "Unknown method");
+   }
+}
+
+static gboolean setup_mock_notifications(GError **error)
+{
+   g_autoptr(GDBusConnection) connection = g_dbus_connection_new_for_address_sync(dbus_address, G_DBUS_CONNECTION_FLAGS_AUTHENTICATION_CLIENT | G_DBUS_CONNECTION_FLAGS_MESSAGE_BUS_CONNECTION, NULL, NULL, error);
```

Comment:
When using GTestDBus, you should be able to use the default session bus connection here.  The g_test_dbus_down() call will ensure that no additional references are held to it on tear down.